group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -65,191 +94,233 @@ public String name() {
+                TopicMetadata metadata = topics.get(topicId);
+                metadata.maybeComputeQuota();
+                addPartitionsToAssignment(metadata, assignment);
             }
+            assignments.put(memberId, new MemberAssignmentImpl(assignment));
+        }
+
+        return new GroupAssignment(assignments);
+    }
 
-            // Step 4 and Step 5
-            // Account for the extra partitions if necessary and increase the required quota by 1.
-            // If remaining > 0 after increasing the required quota, assign the remaining number of partitions from the unassigned partitions list.
-            int unassignedPartitionsListStartPointer = 0;
-            for (MemberWithRemainingAssignments pair : potentiallyUnfilledMembers) {
-                String memberId = pair.memberId;
-                int remaining = pair.remaining;
-                if (numMembersWithExtraPartition > 0) {
-                    remaining++;
-                    numMembersWithExtraPartition--;
-                }
-                if (remaining > 0) {
-                    List<Integer> partitionsToAssign = unassignedPartitionsForTopic
-                        .subList(unassignedPartitionsListStartPointer, unassignedPartitionsListStartPointer + remaining);
-                    unassignedPartitionsListStartPointer += remaining;
-                    newTargetAssignment.computeIfAbsent(memberId, k -> new MemberAssignmentImpl(new HashMap<>()))
-                        .partitions()
-                        .computeIfAbsent(topicId, k -> new HashSet<>())
-                        .addAll(partitionsToAssign);
-                }
+    /**
+     * Sorts the member Ids in the group based on their instance Id if present, otherwise by member Id.
+     * This is done to ensure that the relative ordering of members doesn't change with static members
+     * thus resulting in a sticky assignment.
+     *
+     * @param groupSpec     The group specification containing the member information.
+     * @return a sorted list of member Ids.
+     */
+    private List<String> sortMemberIds(
+        GroupSpec groupSpec
+    ) {
+        List<String> sortedMemberIds = new ArrayList<>(groupSpec.memberIds());
+        Map<String, Optional<String>> instanceIdCache = new HashMap<>();
+
+        for (String memberId : sortedMemberIds) {
+            instanceIdCache.put(memberId, groupSpec.memberSubscription(memberId).instanceId());
+        }
+
+        sortedMemberIds.sort((memberId1, memberId2) -> {
+            Optional<String> instanceId1 = instanceIdCache.get(memberId1);
+            Optional<String> instanceId2 = instanceIdCache.get(memberId2);
+
+            if (instanceId1.isPresent() && instanceId2.isPresent()) {
+                return instanceId1.get().compareTo(instanceId2.get());
+            } else if (instanceId1.isPresent()) {
+                return -1;
+            } else if (instanceId2.isPresent()) {
+                return 1;
```

Comment:
I'd like to better understand this. Are you trying to put all the static members at the beginning (or at the end)? Intuitively, I would have used `instanceId1.orElse(memberId1).compareTo(insteandId2.orElse(memberId2))`.

Reply:
yes all the static members are first and then non-static come after. By putting non-static members  later in the sorted list, we get more flexibility in their partition assignments. Since these members don't have a guarantee of rejoining the group with the same ID, their assignments can be more dynamic. This gives us more room to make sure the static members (with instance IDs) keep their partitions as much as possible, which helps with stickiness and stability.